src/ui/public/share/directives/share.js

```diff
@@ -1,18 +1,146 @@
+        if ($scope.objectId === undefined || $scope.objectId === '') {
+          return;
+        }
+
+        // Remove the query string.
+        const url = $location.absUrl();
+        return url.split('?')[0];
+      };
+
+      function getSnapshotUrl() {
+        const url = $location.absUrl();
+        // Replace hashes with original RISON values.
+        const urlWithStates = unhashUrl(url, getUnhashableStates());
+        return urlWithStates;
+      }
+
+      this.makeUrlEmbeddable = url => {
+        const embedQueryParam = '?embed=true';
+        const urlHasQueryString = url.indexOf('?') !== -1;
+        if (urlHasQueryString) {
+          return url.replace('?', `${embedQueryParam}&`);
+        }
+        return `${url}${embedQueryParam}`;
+      };
+
+      this.makeIframeTag = url => {
+        if (!url) return;
+
+        const embeddableUrl = this.makeUrlEmbeddable(url);
+        return `<iframe src="${embeddableUrl}" height="600" width="800"></iframe>`;
+      };
+
+      this.urls = {
+        original: undefined,
+        snapshot: undefined,
+        shortSnapshot: undefined,
+        shortSnapshotIframe: undefined,
+      };
+
+      this.urlFlags = {
+        shortSnapshot: false,
+        shortSnapshotIframe: false,
+      };
+
+      const updateUrls = () => {
+        this.urls = {
+          original: getOriginalUrl(),
+          snapshot: getSnapshotUrl(),
+          shortSnapshot: undefined,
+          shortSnapshotIframe: undefined,
+        };
+
+        // Whenever the URL changes, reset the Short URLs to regular URLs.
+        this.urlFlags = {
+          shortSnapshot: false,
+          shortSnapshotIframe: false,
+        };
+      };
+
+      // When the URL changes, update the links in the UI.
+      $scope.$watch(() => $location.absUrl(), () => {
+        updateUrls();
+      });
+
+      this.toggleShortSnapshotUrl = () => {
+        this.urlFlags.shortSnapshot = !this.urlFlags.shortSnapshot;
+
+        if (this.urlFlags.shortSnapshot) {
+          urlShortener.shortenUrl(this.urls.snapshot)
+          .then(shortUrl => {
+            this.urls.shortSnapshot = shortUrl;
+          });
+        }
+      };
+
+      this.toggleShortSnapshotIframeUrl = () => {
+        this.urlFlags.shortSnapshotIframe = !this.urlFlags.shortSnapshotIframe;
+
+        if (this.urlFlags.shortSnapshotIframe) {
+          const snapshotIframe = this.makeUrlEmbeddable(this.urls.snapshot);
+          urlShortener.shortenUrl(snapshotIframe)
+          .then(shortUrl => {
+            this.urls.shortSnapshotIframe = shortUrl;
+          });
+        }
+      };
+
+      this.copyToClipboard = selector => {
+        const notify = new Notifier({
+          location: `Share ${$scope.objectType}`,
+        });
+
+        // Select the text to be copied. If the copy fails, the user can easily copy it manually.
+        const copyTextarea = document.querySelector(selector);
```

Comment:
Minor nitpick here: but if you use angular's `$document` service instead of directly accessing the global `document`, it makes testing a good deal easier.